default.nix

```
@@ -1,7 +1,13 @@
 { pkgs ? import <nixpkgs> { } }:
 
 rec {
-  docs = with import ./docs { inherit pkgs; }; {
+  docs = with import ./docs {
+    inherit pkgs;
+    inherit ((pkgs.lib.evalModules {
+      modules = [ ./modules/misc/version.nix ];
+    }).config.home.version)
+      revision nixpkgsRevision;
+  }; {
```

Maybe it's a bit overkill to add an option just for the Nixpkgs revision and evaluate the module. Could perhaps just do something like

```suggestion
  docs = let
    inherit (pkgs.lib) pathIsGitRepo commitIdFromGitRepo trivial;
    revision = let gitRepo = "${toString ./.}/.git";
    in if pathIsGitRepo gitRepo then commitIdFromGitRepo gitRepo else null;
    nixpkgsRevision = trivial.revisionWithDefault null;
    results = import ./docs { inherit pkgs revision nixpkgsRevision; };
  in with results; {
```

?

---

I guess ideally this code will be deprecated at some point in the future anyway if Flakes become non-experimental.

---

So there's a few fiddly things going on here. The root thing driving this is that we want to link the "Declared by:" source information in the manual to the correct revisions of files. We have `home.version.revision` for that, but it tries to figure it out by looking at `.git`, which isn't present when using e.g. flakes or `builtins.fetchGit`. So as of this PR we [inject the correct value](https://github.com/nix-community/home-manager/pull/4215/commits/c066f77ffd52248bc7efcb4c6b1039f8a36eeddb) in `flake.nix`, just like [NixOS does](https://github.com/NixOS/nixpkgs/blob/130e30de7d01ea215870c8d2f4d768444ef810d8/flake.nix#L24-L28).

You'd think that, for nixpkgs, getting the revision is as simple as `lib.trivial.revisionWithDefault null`, as is used in [the default for `nixos.revision`](https://github.com/NixOS/nixpkgs/blob/130e30de7d01ea215870c8d2f4d768444ef810d8/nixos/modules/misc/version.nix#L78-L83), and indeed, that's what I set the default of `home.version.nixpkgsRevision` to as well. Unfortunately it's not that simple: the injection in Nixpkgs' flake is only done for NixOS systems, not the Nixpkgs library itself. So we [forward the value from the system](https://github.com/nix-community/home-manager/pull/4215/commits/832cca945cdcd02b81d535780c49e9f4aa126b19) in the NixOS and nix-darwin modules, as they already go to efforts to get the correct revision when using flakes. Going forward, I intend to send a PR to Nixpkgs to override `lib.trivial.revisionWithDefault` in the flake, which would allow simplifying the logic everywhere else in the ecosystem.

It's a bit of a fuss, but it's only some 15 lines of code, and the Nixpkgs revision could be useful in other ways (e.g. accurately labelling generations with their corresponding precise Nixpkgs version like NixOS and nix-darwin do). That said, right now it's only used to link the declaration of global Nixpkgs module system options, so it's not exactly vital.

Even without `nixpkgsRevision`, however, we would need to do something like this in `default.nix` to get the more important links to the Home Manager repository right (especially since it's used to render the manual in CI). The diff you suggest would work, but would lead to `default.nix` duplicating the logic in `modules/misc/version.nix`; I figured the slightly awkward contortion here beats two sources of truth that could diverge. The other alternative I explored was setting the defaults in a new `version.nix` that both `default.nix` and `modules/misc/version.nix` could import, but it didn't feel worth making `modules/misc/version.nix` more awkward just to avoid calling `lib.evalModules` in `default.nix`. Open to diverging opinions here though :)

---

Had a think about this. It might be a good idea to use branch names rather than exact commits anyway, to reduce manual rebuilds, and hopefully the drift will be pretty minor. The only thing is that I don't know if there's a method hooked up to determine whether we're on a release branch (so we should link to `release-${release}`) or unstable (so we should link to `master`)? A simple boolean would be enough but it'd have to be flipped on every branch-off.

---

There isn't, but I have suggested that in https://github.com/nix-community/home-manager/pull/4110#issue-1762201781 to smooth out the release check around branch-off.

---

Thanks for the pointer. If a boolean signalling whether the branch is `master` would be useful for other things too then I'd be happy to introduce it in this PR and use it to determine branches to link to.

---

I think that would make sense. The release flow would be 1. branch off, 2. set `stable = true` in the new release branch, 3. bump the release in `master`

---

Sorry for the delay. Good points, no need to do any change with respect to the `evalModules`.

I do agree that we need to change the links to be to the branch name rather than the actual commit hash. I hadn't noticed this so it was good you brought it up, I think this is pretty important to avoid the rebuilds. Fortunately, I think this is pretty simple to fix, just hard-code it in the link, we can have it as a manual release task to switch the link to `release-{whatever}`. It would be a very quick thing to do twice per year so no need to over-complicate it.

---

I started implementing this before seeing this comment; adding the boolean was very easy and ended up being much less code than the fuss it was to deal with Nixpkgs revisions, as well as getting rid of the `evalModules`, so I've kept it in the latest push. I think @ncfavier's points about reducing the number of things required for branch-off rather than adding more steps are good, and it would be easy to gate or relax (e.g. to `versionOlder`) the release mismatch check based on this, so I think it pays for itself. (Plus nix-darwin is going to need something like this too if we add release branches, so it was useful practice.) But if you still think it's too much I can get rid of it.